coderd/notifications/metrics.go

```diff
@@ -0,0 +1,71 @@
+package notifications
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type Metrics struct {
+	DispatchedCount  *prometheus.CounterVec
+	TempFailureCount *prometheus.CounterVec
+	PermFailureCount *prometheus.CounterVec
+	RetryCount       *prometheus.CounterVec
+
+	QueuedSeconds *prometheus.HistogramVec
+
+	DispatcherSendSeconds *prometheus.HistogramVec
+
+	PendingUpdates prometheus.Gauge
+}
+
+const (
+	ns        = "coderd"
+	subsystem = "notifications"
+
+	LabelMethod     = "method"
+	LabelTemplateID = "template_id"
+)
+
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	return &Metrics{
+		DispatchedCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "dispatched_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications successfully dispatched.",
+		}, []string{LabelMethod, LabelTemplateID}),
+		TempFailureCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "temporary_failures_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications which failed but have retry attempts remaining.",
+		}, []string{LabelMethod, LabelTemplateID}),
+		PermFailureCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "permanent_failures_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications which failed and have exceeded their retry attempts.",
+		}, []string{LabelMethod, LabelTemplateID}),
```

Comment:
DispatchCount, TempFailureCount, and PermFailureCount might make sense as a single metric with different "disposition" (or "result") labels.  The help text refers to them as "notifications", but I think that's misleading.  What we are counting is "notification attempts" or "dispatch attempts", which might individually end up in success or some kind of failure.  In particular, in the case of failure and retry, the number of delivery attempts can exceed the number of notifications, so it's important to make the distinction.

Prometheus guidelines say that something should be a single metric with labels (vs multiple metrics) when the sum or average of the label is sensible and coherent. Filtering by method and summing across disposition would be useful, for example, to track the number of web requests we are making via Webhook delivery.